Add more documentation (#36)
DataBiosphere/wdl-conformance-tests: README.md

                        Arguments to pass into cromwell. Ex: --cromwell-args="--options=[OPTIONS]"
  --id ID               Specify WDL tests by ID.
  --repeat REPEAT       Specify how many times to run each test.
  --jobstore-path JOBSTORE_PATH, -j JOBSTORE_PATH
                        Specify the PARENT directory for the jobstores to be created in.
  --progress            Print the progress of the test suite as it runs.

```
Invoking `run.py` with no options will result in the entire test suite being ran, which can take a long time.
Including `--progress` is recommended when running over a long period of time.

Certain tests can be specified with `--id`, `--tags`, and `--numbers`:
```commandline
python run.py --number 1 --id md5 --tags stderr --version 1.0 --runner toil-wdl-runner
Testing runner toil-wdl-runner on WDL versions: 1.0

=== REPORT ===

1: SUCCEEDED: Standard Lib: Basic test for stderr()
Iteration: 1
40: SUCCEEDED: Legacy test for stderr_as_output
Iteration: 1
68: SUCCEEDED: MD5 test
Iteration: 1
3 tests run, 3 succeeded, 0 failed, 0 skipped
```
`--repeat` specifies how many times to run each test. `--threads` allows multiple tests to run simultaneously;

> this should be set to the number of processors and not threads in a computer if running with `--time` or with the performance testing script.

Author: There's no machinery currently to reserve certain threads. Most if not all of the tests only allocate 1 cpu core, but I don't think the behavior of how to run the workflow/task section on what cores is strictly defined (as in, if task allocates 1 cpu, it might run on a separate thread than the workflow). I think the best way is to tweak the documentation to imply that the maximum amount of tests to run is the number of threads excluding hyperthreading, and that consistent timings may not be fully guaranteed if a runner uses more threads as specified and once all threads are saturated.

  versions: ["draft-2", "1.0", "1.1"] # specify versions to run with
  id: example_test_id # unique ID of test, no 2 tests should have the same id
  tags: ["examples", "behavior"] # specify tags, these don't need to be unique
  inputs:
    dir: tests/example_files # path to directory where test is, can be an absolute or relative path (to run.py)
    wdl: example.wdl # wdl file name
    json: example.json # json file name
  outputs:
    exampleWf.outputVar: # output name, should be workflowName.outputVariable
      type: Boolean # expected output type
      value: True # expected output value
```
The test files then should be under the `tests/example_files` directory and be named `example.wdl` and `example.json`.
Each expected output should be specified as `wfName.varName`. For example, `exampleWf.outputVar` is the specifier if the WDL is this:
```wdl
workflow exampleWf {
    output {
        Boolean outputVar = true
    }
}
```
Only one WDL file is necessary to run in all versions, as long as formatting is followed as [stated below](#test-formatting).
## Test formatting
Additionally, in order to keep a single WDL file per each test, the same test for other versions are generated.

> There are two ways for generating these tests: [automatic](#automatic-formatting) or [manual](#manual-formatting).

Author: Maybe version conversion is a better name

The test runner uses a single WDL file for each test, written for a single WDL version, to run the test across all applicable WDL versions. This is accomplished by rewriting that one WDL file at runtime to produce generated WDL files targeting the other WDL versions.

Test can use either [automatic version conversion](#automatic-version-conversion) or [manual version conversion](#manual-version-conversion) to be runnable across multiple WDL versions from a single file. If a test uses neither, it must be restricted to a single WDL version.
### Automatic version conversion
Different WDL versions (`draft-2`, `1.0`, `1.1`) all have slightly different syntax. To deal with this, the runner will 
automatically convert from one version to another before running a WDL test. As long as the format of a WDL file 
follows these format conventions, the conversion will work properly:

Any input section in WDL code must follow the format:
```wdl
input {
    ...
}
```
Any command section must follow the format:
```wdl
command {
    ...
} # line with isolated closing brace indicates end of command section
```
Any command section can also use triple angle braces:
```wdl
command <<<
    ...
>>>
```

> These sections should not have other WDL code spill over or have unnecessary newlines in between the section declaration.

Reviewer: ```suggestion
These sections should not have other WDL code on the same line, or have unnecessary newlines in between the section declaration and its block.
```

```commandline
python patch.py --version 1.1 --directory tests/basic_stdout
```
The directory `tests/basic_stdout` is the directory that holds all the WDL files written in all versions. The specified version will be the base file to make patch files from.
`--remove` can be provided to delete all other WDL files that aren't the base WDL file, and `--rename` can be used to rename the base WDL file to the same name as the directory.

For example, if `tests/basic_stdout` has 3 WDL files:
```commandline
tests/basic_stdout
├── basic_stdout.json
├── basic_stdout_version_1.0.wdl
├── basic_stdout_version_1.1.wdl
└── basic_stdout_version_draft2.wdl
```
After `python patch.py --version 1.1 --directory tests/basic_stdout` is called, then `patch.py` will use `basic_stdout_version_1.1.wdl` as the base file for patches, and generate 2 patch files to convert from WDL 1.1 to 1.0 and WDL 1.1 to draft-2.
```commandline
tests/basic_stdout
├── ...
├── version_1.0.patch
└── version_draft-2.patch
```
These patch files will be used to create the proper versioned WDL file at runtime (This will take priority over the automatic WDL version conversion).

## Running Performance Tests

> The default runner `run.py` only runs tests and outputs them onto the commandline.

Reviewer: ```suggestion
The default runner `run.py` only runs tests and reports on their status in the terminal.
```

To store the runtime of tests (and eventually graph them), the script `run_performance.py` should be used:
```commandline
python run_performance.py --runners toil-wdl-runner,miniwdl --output performance_output.csv --progress
```

> The output will always be in CSV format. `--runners` can specify runners to run with in a comma separated list, and `--all-runners` is a shortcut to specify all of them.

Reviewer: ```suggestion
The output will always be in CSV format. `--runners` can specify a comma separated list of runners to test, and `--all-runners` is a shortcut to specify all of them.
```

### Performance Testing Options

> [All options](#options) from the normal `run.py` script apply too. For example, if `--id stdout` is provided, then only the `stdout` test will be measured.

Reviewer: ```suggestion
[All options](#options) from the normal `run.py` script are also available in `run_performance.py`. For example, if `--id stdout` is provided, then only the `stdout` test will be measured.
```

```commandline
Arguments for graphing runtimes of WDL tests:
  --from-file FILE, -f FILE
                        Specify a csv file to read from.
  --display-num DISPLAY_NUM, -d DISPLAY_NUM
                        Specify the number of tests to display per graph.
  --display-all, -a     Display all tests on a single graph. Overrides --display-num.
  --ignore-runner IGNORE_RUNNER
                        Specify a runner(s) to ignore in the graph output.
  --precision PRECISION
                        Specify the precision when outputting float values. Ex: Default=0 will result in 1 for float value 1.4...
  --no-labels           Specify to not display extra labels on the graph.
  --graph-type GRAPH_TYPE
  --conformance-file [CONFORMANCE_FILE]
                        Specify the conformance file to read from. This will specify whether to grab/graph tests by conformance file or by CSV file test IDs only. Specifying this will make the graph accept -n, -t, -id and
                        other related arguments.
```
By default, separate graphs will be created for every 30 tests.

> This can be controlled with `--display-num`. This overrides the number of tests to display per graph, and `--display-all` forces all tests in a single graph.

Reviewer: ```suggestion
This can be controlled with `--display-num`, which overrides the number of tests to display per graph. `--display-all` forces all tests in a single graph.
```

`--graph-type` specifies the type of graph to create. The supported options are `box` and `bar`.

> `box` will create a traditional box plot and `bar` will create a bar graph with error bars representing standard deviation. 

Reviewer: It's always important to say what the error bars are! Good job!

### Writing Graphs to a File

> The graphs can be written to a file(s).

Reviewer: ```suggestion
The graphs can be written to disk by specifying a (base) filename.
```

For example, to output a graph with all tests in a PNG with a size of 40 inches by 10 inches (width,height):
```commandline
python create_graph.py -f performance_output.csv --output performance_graph.png --dimensions 40,10 -a
```
If `--display-all` is not specified and there are more tests to graph than `--display-num`, then multiple graph files will be created with the format `[output_file][iteration].[fileext]`.

For example, calling `python create_graph.py -f performance_output.csv --output performance_graph.png` with 90 total tests to graph will result in:
```commandline
.
├── performance_graph1.png
├── performance_graph2.png
└── performance_graph3.png
```
If trying to put all the tests into a single graph, it is likely that dimensions should be provided to make all tests fit.

> All output formats supported by Matplotlib are also supported (ex: pdf, svg).

Reviewer: ```suggestion
All output formats implemented by Matplotlib (such as `pdf` and `svg`) are supported.
```
